drivers: nrf: Add power management to nrf SPIM driver.

Mierunski opened the pull request:
Add power management functions to nrf SPIM driver.

Fix multi instance issues in TWI and TWIM drivers.
Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski committed:
drivers: nrf: Add power management to nrf SPIM driver.

Add power management functions to nrf SPIM driver.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

tbursztyka commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -344,10 +373,22 @@ static int init_spim(struct device *dev, const nrfx_spim_config_t *config)
 	};								       \
-	DEVICE_AND_API_INIT(spi_##idx,					       \
+	DEVICE_DEFINE(spi_##idx,					       \
 			    DT_NORDIC_NRF_SPI_SPI_##idx##_LABEL,	       \
```
indentation of parameters should be right below last '(' + 1 space

tagunil commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -290,13 +291,14 @@ static void event_handler(const nrfx_spim_evt_t *p_event, void *p_context)
 	/* This sets only default values of frequency, mode and bit order.
 	 * The proper ones are set in configure() when a transfer is started.
 	 */
 	nrfx_err_t result = nrfx_spim_init(&get_dev_config(dev)->spim,
```
&dev_config->spim maybe?

Mierunski committed:
drivers: nrf: Add power management to nrf SPI driver.

Add power management to nrf SPI driver.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

anangl commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -310,6 +311,43 @@ static int init_spim(struct device *dev, const nrfx_spim_config_t *config)
+
+		if (new_state != current_state) {
+			if (new_state == DEVICE_PM_ACTIVE_STATE) {
+				nrfx_spim_init(&dev_config->spim,
```
We should use consistent approach in these drivers, i.e. use either `nrfx_spi[m]_init` or `init_spi[m]` in both. I think the latter option will be better as it will be also consistent with what we'll have in TWI/TWIM drivers.

Mierunski commented:
@anangl @carlescufi ping

anangl commented on drivers/spi/spi_nrfx_spi.c:
```diff
@@ -271,23 +272,49 @@ static int init_spi(struct device *dev, const nrfx_spi_config_t *config)
+static int spi_nrfx_pm_control(struct device *dev, u32_t ctrl_command,
+				void *context, device_pm_cb cb, void *arg)
+{
+	static u32_t current_state = DEVICE_PM_ACTIVE_STATE;
```
This cannot be a common variable for all instances, like it is now. Each instance needs to store its own state separately. Same applies to SPIM (and to TWI and TWIM that have already been merged, unfortunately).

anangl commented on drivers/spi/spi_nrfx_spi.c:
```diff
@@ -271,23 +272,49 @@ static int init_spi(struct device *dev, const nrfx_spi_config_t *config)
+			switch (new_state) {
+			case DEVICE_PM_ACTIVE_STATE:
+				init_spi(dev);
+				configure(dev, get_dev_data(dev)->ctx.config);
```
This is not safe. Before `tranceive` (and consequently `configure`) is called for the first time, `ctx.config` is `NULL`. And nothing prevents `spi_nrfx_pm_control` from being called at that time. You should change this line to:
```
get_dev_data(dev)->ctx.config = NULL;
```
preferably with a comment explaining that it is done to force reconfiguration of the peripheral before the next transfer.

anangl commented on drivers/spi/spi_nrfx_spi.c:
```diff
@@ -271,23 +272,49 @@ static int init_spi(struct device *dev, const nrfx_spi_config_t *config)
+				configure(dev, get_dev_data(dev)->ctx.config);
+				break;
+
+			default:
```
Shouldn't we return `-ENOTSUP` for `DEVICE_PM_FORCE_SUSPEND_STATE` (and any other unknown state that might be added in some future) instead of pretending that we handle it properly?

Mierunski committed:
drivers: nrf: Fix PM for TWI and TWIM in case of multi instance

TWI and TWIM used single static variable for multiple instances.
It would cause problems in case of multiple instances of peripheral.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

anangl commented on drivers/i2c/i2c_nrfx_twi.c:
```diff
@@ -157,28 +163,34 @@ static int init_twi(struct device *dev)
 				break;
+
+			default:
+				return -ENOTSUP;
```
The value needs to be passed as the second parameter to the `cb` function. At least that's what other drivers do. This parameter is not well/at all documented in `device.h` (its name just says `status`), so I think we should just follow the others.

anangl replied:
Maybe you could use the pattern from [the PWM driver](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/pwm/pwm_nrfx.c#L322)?

anangl commented on drivers/spi/spi_nrfx_spi.c:
```diff
@@ -266,28 +270,63 @@ static int init_spi(struct device *dev, const nrfx_spi_config_t *config)
+			default:
+				ret = -ENOTSUP;
+			}
+			get_dev_data(dev)->pm_state = new_state;
```
`if (!ret) {`

Mierunski commented:
@carlescufi ping
